File: leetcodes/98. isValidBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 *
 *   2 1 3
 *   input: val, max, min
 *   2  check left -> if 1 < 2
 *                    if min > 2 return false
 *      check right -> if 3 < 2
 *                    if max < 2 return false
 *
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
  //recursive
  //O(n) time
  if (!root) {
    return true;
  }
  function helper(root, small = -Infinity, big = +Infinity) {
    if (!root) {
      return true;
    }
    if (root.val <= small || root.val >= big) {
      return false;
    }
    return (
      helper(root.left, small, root.val) && helper(root.right, root.val, big)
    );
  }
  return helper(root);
};
//iterative

File: leetcodes/98. isValidBST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True

        def helper(node, min, max):
            if not node:
                return True
            if (node.val >= max or node.val <= min):
                return False
            return helper(node.right, node.val, max) and helper(node.left, min, node.val)
        return helper(root, float('-inf'), float('inf'))
